### PR TITLE
Add `sort_samples` argument to `join`

### LINKED ISSUE #467: Different ordering in `join`, apart from the default lexicographical ordering, should be an option
Say we are collating 2 TensorMaps corresponding to different samples in order to form a minibatch, for instance using `metatensor.learn.data.group_and_join`, the default collate function of `metatensor.learn.data.DataLoader`. `join` will lexicographically sort the samples metadata, but in often it is useful to retain the original ordering. This is for instance important when joining TensorMap descriptors / targets but working with an ordered list of systems as input.

Perhaps `join` should come with user-controllable options for how to sort the (meta)data in the resulting TensorMap.

**MWE**

Joining two TensorMaps that are identical, except reverse ordered sample indices `1` and `0`, respectively:

```py
import numpy as np
import metatensor
from metatensor import Labels, TensorBlock, TensorMap

tensors = [
    TensorMap(  # sample "1"
        keys=Labels(names=["a"], values=np.array([[0]])),
        blocks=[
            TensorBlock(
                values=np.array([0]).reshape(-1, 1),
                samples=Labels(names=["s"], values=np.array([1]).reshape(-1, 1)),
                components=[],
                properties=Labels(names=["p"], values=np.array([0]).reshape(-1, 1)),
            )
        ]
    ),
    TensorMap(  # sample "0"
        keys=Labels(names=["a"], values=np.array([[0]])),
        blocks=[
            TensorBlock(
                values=np.array([0]).reshape(-1, 1),
                samples=Labels(names=["s"], values=np.array([0]).reshape(-1, 1)),
                components=[],
                properties=Labels(names=["p"], values=np.array([0]).reshape(-1, 1)),
            )
        ]
    ),
]

joined_tensor = metatensor.join(tensors, axis="samples", remove_tensor_name=True)
```

The order of the samples of the joined TensorMap are reversed:

```py
joined_tensor.block(0).samples
>>> Labels(
    s
    0
    1
)
```

This has implications for `metatensor-models` - tagging @frostedoyster, @agoscinski and @PicoCentauri involved in this.

### PR DESCRIPTION
**Pull request by PicoCentauri**
Fixes #467

Thanks for the detailed examples @jwa7.

Maybe we should update the Changelog @Luthaf. This is a breaking change.

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [x] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [x] CHANGELOG updated with public API or any other important changes?

----
📚 Documentation preview 📚: https://metatensor--476.org.readthedocs.build/en/476/

### CODE REVIEW

**Comment by Luthaf** [comment]
Yes, this should be added to the changelog

**Review by jwa7 — APPROVED** [review]
Looks good, thanks @PicoCentauri ! I've updated the change log, happy for you to squash and merge once the CI has passed :)

File: python/metatensor-operations/CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,9 @@ follows [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### Added
 
+- `join()` operation now includes the `sort_samples` argument to control the
+  sorting of samples in the output joined TensorMap. This defaults to False, opposite
+  to the previous default behaviour.
```
[the remark this answers is not in the document]
Comment (jwa7) [reply python/metatensor-operations/CHANGELOG.md:24]: Good point! Done.

[the remark this answers is not in the document]
Comment (PicoCentauri) [reply python/metatensor-operations/CHANGELOG.md:24]: It was at changed before but I thought maybe Added is better xD

**Pull request merged**